lib/sourcebit.js

```diff
@@ -226,13 +231,29 @@ class Sourcebit {
                     return data;
                 }
 
-                return plugin.transform({
-                    data,
+                const { __diff, ...currentDataForPlugin } = data;
+                const previousData = this.dataForPlugin[index] || initialData;
+                const diffs = Object.keys(currentDataForPlugin).reduce((diffs, dataBucketKey) => {
+                    return {
+                        ...diffs,
+                        [dataBucketKey]: generateDiff(previousData[dataBucketKey], currentDataForPlugin[dataBucketKey]) || []
+                    };
+                }, {});
+
+                this.dataForPlugin[index] = currentDataForPlugin;
+
+                const transformedData = await plugin.transform({
+                    data: {
+                        ...data,
+                        __diff: diffs
+                    },
                     debug: this.getDebugMethodForPlugin(pluginName),
                     getPluginContext: () => contextSnapshot[pluginName] || {},
                     log: this.logFromPlugin.bind(this),
                     options: this.parsePluginOptions(plugin, pluginBlock.options)
                 });
+
+                return transformedData;
```

Comment:
It looks like the data is passed to plugin `transform` functions by reference and also stored by reference in `dataForPlugin`. This means that if one plugin mutates the data, it will be changed in `dataForPlugin` as well, and therefore affect the next diff iteration.

Example:

```
data = {
  objects: [{"foo": "bar"}]
}
=> currentDataForPlugin = {objects: [{"foo": "bar"}]} // same object
=> this.dataForPlugin[index] = {objects: [{"foo": "bar"}]} // same object
plugin.transform({data: {...data}})
```
Plugin code
```
transform({data, ...}) {
  data.objects.foo.bar = "changed";
}
```
This changes the value inside the `data` passed in transform but also the value inside `this.dataForPlugin[index]`.

Next time transform runs, the `previousData` for that plugin will have the value `changed` and not the original value.

The data passed to `transform` function (or the data saved into `dataForPlugin`) should be deeply cloned. Or better yet, deeply freeze the data before passing it to plugin `transform` (or use [immutable.js](https://immutable-js.github.io/immutable-js/)), requiring plugin developers to always return new objects by leveraging functional development and use of pure functions.

Reply:
That's true. I did consider this when I was working on the implementation, but I worry that by freezing/cloning the data we'll be adding too much effort. We already clone the entire data object on every transform, now we're talking about doing it also for every plugin, on every transform.

I mean, we can do it if you feel that it's important, but it seems that we've been on the fence about the importance of the whole diff object concept, so I'm a bit weary about making it considerably more expensive. On the other hand, there's the argument that if we're doing it, we might as well do it right! 😄 

What do you think?

Reply:
"On every transform" you mean every transform of every plugin? Or every global transform run?
If this is for global transform run, then the impact is low.
But if the diff is run for every plugin, then my question is  do we need to run it for every plugin? If I understand it correctly, the diff needs to be run once before iterating over the plugins and running their `transform` functions. Because what we want is the diff between the whole data, before running any plugins.

Reply:
Anyway, we should check if deep-freeze has any significant performance impact. If not, then I would safely use it.